lib/interactor.rb

```diff
@@ -112,8 +112,16 @@ def initialize(context = {})
   #
   # Returns nothing.
   def run
-    run!
-  rescue Failure
+    catch(:early_return) do
+      with_hooks do
+        call
+        context.called!(self)
+      end
+    end
+    context.rollback! if context.failure?
+  rescue
+    context.rollback!
+    raise
```

Comment:
I'd like to see if we could preserve how the primary functionality lived  in `run!` rather than in `run`.